src/realm/set.hpp

```diff
@@ -0,0 +1,455 @@
+    bool operator()(const Mixed& a, const Mixed& b) const noexcept
+    {
+        // CAUTION: This routine is technically part of the file format, because
+        // it determines the storage order of Set elements.
+
+        if (a.is_null() != b.is_null()) {
+            return false;
+        }
+        else if (a.is_null()) {
+            return true;
+        }
+
+        if (a.get_type() != b.get_type()) {
+            return false;
+        }
+
+        switch (a.get_type()) {
+            case type_Int:
+                return a.get<int64_t>() == b.get<int64_t>();
+            case type_Bool:
+                return a.get<bool>() == b.get<bool>();
+            case type_String:
+                return a.get<StringData>() == b.get<StringData>();
+            case type_Binary:
+                return a.get<BinaryData>() == b.get<BinaryData>();
+            case type_Timestamp:
+                return a.get<Timestamp>() == b.get<Timestamp>();
+            case type_Float:
+                return a.get<float>() == b.get<float>();
+            case type_Double:
+                return a.get<double>() == b.get<double>();
+            case type_Decimal:
+                return a.get<Decimal128>() == b.get<Decimal128>();
+            case type_ObjectId:
+                return a.get<ObjectId>() == b.get<ObjectId>();
+            case type_TypedLink:
+                return a.get<ObjLink>() == b.get<ObjLink>();
+            case type_OldTable:
+                [[fallthrough]];
+            case type_Mixed:
+                [[fallthrough]];
+            case type_OldDateTime:
+                [[fallthrough]];
+            case type_Link:
+                [[fallthrough]];
+            case type_LinkList:
+                REALM_TERMINATE("Invalid Mixed payload in Set.");
+        }
+        return false;
+    }
+};
+
+template <class T>
+inline Set<T>::Set(const Obj& obj, ColKey col_key)
+    : Collection<T, SetBase>(obj, col_key)
+{
+    if (m_obj) {
+        this->init_from_parent();
+    }
+}
+
+template <typename U>
+Set<U> Obj::get_set(ColKey col_key) const
+{
+    return Set<U>(*this, col_key);
+}
+
+template <class T>
+size_t Set<T>::find(T value) const
+{
+    auto b = this->begin();
+    auto e = this->end();
+    auto it = std::lower_bound(b, e, value, SetElementLessThan<T>{});
+    if (it != e && SetElementEquals<T>{}(*it, value)) {
+        return it.index();
+    }
+    return npos;
+}
+
+template <class T>
+size_t Set<T>::insert(T value)
+{
+    REALM_ASSERT_DEBUG(!update_if_needed());
+
+    ensure_created();
+    this->ensure_writeable();
+    auto b = this->begin();
+    auto e = this->end();
+    auto it = std::lower_bound(b, e, value, SetElementLessThan<T>{});
+
+    if (it != e && SetElementEquals<T>{}(*it, value)) {
+        return it.index();
+    }
+
+    if (Replication* repl = m_obj.get_replication()) {
+        // FIXME: We should emit an instruction regardless of element presence for the purposes of conflict
```

Comment:
It's tricky for Core replication, because there we only want to emit an index of the inserted value, but if we create duplicate insertions, the size of the set can no longer be computed by looking at just the insert instructions.

We can add a bool for it later, but I would rather defer the discussion, because it's a bit of an edge case.